File: src-emulator/com/jediterm/emulator/TerminalDataStream.java
package com.jediterm.emulator;

import java.io.IOException;

/**
 * @author traff
 */
public interface TerminalDataStream {
  char getChar() throws IOException;

  void pushChar(char c) throws IOException;

  CharacterUtils.CharArraySlice advanceThroughASCII(int maxChars) throws IOException;

  void sendBytes(byte[] response) throws IOException;

  void pushBackBuffer(char[] bytes, int i) throws IOException;

  class EOF extends IOException {
    public EOF() {
      super("EOF: There is no more data or connection is lost");
    }
  }
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/TerminalDataStream.kt
package com.jediterm.terminal

import java.io.IOException

/**
 * Represents data communication interface for terminal.
 * It allows to [.getChar] by one and [.pushChar] back as well as requesting a chunk of plain ASCII
 * characters ([.readNonControlCharacters] - for faster processing from buffer in the size <=**maxChars**).
 *
 *
 * @author traff
 */
interface TerminalDataStream {
    @get:Throws(IOException::class)
    val char: Char

    @Throws(IOException::class)
    fun pushChar(c: Char)

    @Throws(IOException::class)
    fun readNonControlCharacters(maxChars: Int): String?

    @Throws(IOException::class)
    fun pushBackBuffer(bytes: CharArray?, length: Int)

    val isEmpty: Boolean

    class EOF : IOException("EOF: There is no more data or connection is lost")
}
